// react-router-modal/lib/modal_route.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; // eslint-disable-line no-unused-vars


var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactRouterDom = require('react-router-dom');

var _modal = require('./modal');

var _modal2 = _interopRequireDefault(_modal);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function getStackOrder(match) {
  // order routed modals behind any directly instantiated modals
  return match.url.length - 10000;
}

/**
* A react-router Route that shows a modal when the location pathname matches.
*
* @param {Object} props
* @param {String} props.path path to match
* @param {Boolean} props.exact If set, only show modal if route exactly matches path.
* @param {String} props.parentPath path to navigate to when backdrop is clicked
*
* @param {String} props.className class name to apply to modal container
* @param {Children} props.children modal content can be specified as chld elements
* @param {ReactElement} props.component modal content can be specified as a component type
* @param {Object} props.props Props to be passed to the react component specified by the component property.
*
*
* When the route matches, the modal is shown.
* If multiple routes match, the modals will be stacked based on the length of the path that is matched.
*
*/
function ModalRoute(_ref) {
  var path = _ref.path,
      parentPath = _ref.parentPath,
      className = _ref.className,
      children = _ref.children,
      component = _ref.component,
      exact = _ref.exact,
      props = _ref.props,
      match = _ref.match,
      history = _ref.history;


  var navToParent = function navToParent() {
    if (typeof parentPath === 'function') {
      return history.push(parentPath(match));
    }
    history.push(parentPath || match.url);
  };

  return _react2.default.createElement(_reactRouterDom.Route, { path: path, exact: exact, render: function render(_ref2) {
      var match = _ref2.match,
          location = _ref2.location,
          history = _ref2.history;
      return _react2.default.createElement(_modal2.default, {
        component: component,
        children: children,
        props: _extends({}, props, { match: match, location: location, history: history }),
        className: className,
        stackOrder: getStackOrder(match),
        onBackdropClick: navToParent
      });
    } });
}

exports.default = (0, _reactRouterDom.withRouter)(ModalRoute);
module.exports = exports['default'];
